<a href="https://colab.research.google.com/github/suhanisingh5617-ai/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/suhanisingh5617-ai/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I'm going with the CTR / Engagement Opportunity Scoring lane. In Week 1, I already dug into how click-through rate varies across content type and position tier, and this lane felt like a natural next step — turning that exploration into something more structured: a ranked score that actually points to where the opportunity is.


In [ ]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/suhanisingh5617-ai/flyrank-ml-internship"
REPO_DIR = "flyrank-ml-internship"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)

import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

df['position_tier'] = pd.cut(df['avg_position'], bins=[0,3,10,20,100], labels=['top3','4-10','11-20','20+'])
result = df.groupby(['position_tier','content_type'])['ctr'].mean().sort_values()
print(result)

position_tier  content_type      
top3           comparison article    0.000000
20+            comparison article    0.098846
4-10           comparison article    0.131259
11-20          comparison article    0.147374
20+            keyword article       0.169078
11-20          keyword article       0.280972
4-10           keyword article       0.447746
top3           keyword article       1.251822
11-20          feedly article        2.119669
4-10           feedly article        3.350570
20+            feedly article        3.404386
top3           feedly article        9.872680
Name: ctr, dtype: float64


/tmp/ipykernel_2463/3861267477.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  result = df.groupby(['position_tier','content_type'])['ctr'].mean().sort_values()


## 2. The question: decision, action, cost of a wrong call

The question I want to answer is: which pages are underperforming on CTR relative to where they rank, and how much room is there to improve? The decision this feeds into is which pages a content or SEO team should prioritize for optimization. The action would be flagging those pages for a title/meta rewrite or content refresh. If the score gets it wrong, the cost isn't huge but it's real — the team could spend time optimizing a page that was already fine, or worse, miss a high-traffic page that's genuinely underperforming.

In [ ]:
underperforming = df[df['ctr'] < df.groupby('position_tier')['ctr'].transform('median')]
print("Pages underperforming vs their tier median:", len(underperforming))

Pages underperforming vs their tier median: 9400


/tmp/ipykernel_2463/269573831.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  underperforming = df[df['ctr'] < df.groupby('position_tier')['ctr'].transform('median')]


## 3. Quick look at the data (2-3 real numbers)

From my Week 1 discovery, I already have some evidence worth citing: in the top3 position tier, "feedly article" content had a CTR of 9.87, while "comparison article" content in the same tier had a CTR close to 0.00. That's a huge gap for pages ranking in the same spot, which suggests content type alone is doing a lot of the work here — worth digging into further.

In [ ]:
top3_feedly = df[(df['position_tier']=='top3') & (df['content_type']=='feedly article')]['ctr'].mean()
top3_comparison = df[(df['position_tier']=='top3') & (df['content_type']=='comparison article')]['ctr'].mean()

print("Top3 tier — feedly article CTR:", round(top3_feedly, 3))
print("Top3 tier — comparison article CTR:", round(top3_comparison, 3))
print("Gap:", round(top3_feedly - top3_comparison, 3))

Top3 tier — feedly article CTR: 9.873
Top3 tier — comparison article CTR: 0.0
Gap: 9.873


## 4. Careful words: what I can and can't claim

I can say that CTR differences across content type and position are observed and directional in this dataset. I can't say I've proven why users click less on certain content types, or that fixing this will definitely raise CTR — that would need testing, not just observation.

In [ ]:
print("Total pages in dataset:", df.shape[0])

Total pages in dataset: 30000


## Self-check

I've picked a lane that builds directly on what I already explored, named a clear decision and action this could support, backed it with real numbers from the starter data, and kept my claims to what the data actually shows rather than overstating them.